src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala

```diff
@@ -101,38 +93,40 @@ class OrderedRVDPartitioner(
 
 
   def withKType(newPartitionKey: Array[String], newKType: TStruct): OrderedRVDPartitioner = {
-    val (newPKType, _) = newKType.select(newPartitionKey)
-    val newRangeBounds = new UnsafeIndexedSeq(TArray(TInterval(newPKType)), rangeBounds.region, rangeBounds.aoff)
-    val newPart = new OrderedRVDPartitioner(newPartitionKey, newKType, newRangeBounds)
+    val newPart = new OrderedRVDPartitioner(newPartitionKey, newKType, rangeBounds)
     assert(newPart.pkType.types.sameElements(pkType.types))
     newPart
   }
 
   def copy(numPartitions: Int = numPartitions, partitionKey: Array[String] = partitionKey,
-    kType: TStruct = kType, rangeBounds: UnsafeIndexedSeq = rangeBounds): OrderedRVDPartitioner = {
+    kType: TStruct = kType, rangeBounds: IndexedSeq[Interval] = rangeBounds): OrderedRVDPartitioner = {
     new OrderedRVDPartitioner(partitionKey, kType, rangeBounds)
   }
 
   // FIXME Make work if newRange has different point type than pkType
   def enlargeToRange(newRange: Interval): OrderedRVDPartitioner = {
-    val newStart = pkType.ordering.min(range.start, newRange.start)
-    val newEnd = pkType.ordering.max(range.end, newRange.end)
-    val newRangeBounds = rangeBounds.toArray
-    newRangeBounds(0) = newRangeBounds(0).asInstanceOf[Interval]
-      .copy(start = newStart, includesStart = true)
-    newRangeBounds(newRangeBounds.length - 1) = newRangeBounds(newRangeBounds.length - 1)
-      .asInstanceOf[Interval].copy(end = newEnd, includesEnd = true)
-    copy(rangeBounds = UnsafeIndexedSeq(rangeBoundsType, newRangeBounds))
+    val newStart = pkType.ordering.min(range.start, Annotation.copy(pkType, newRange.start))
+    val newEnd = pkType.ordering.max(range.end, Annotation.copy(pkType, newRange.end))
+    val newRangeBounds =
+      rangeBounds match {
+        case IndexedSeq(x) => IndexedSeq(Interval(newStart, newEnd, true, true))
+        case IndexedSeq(x1, x2) =>
+          IndexedSeq(x1.copy(start = newStart, includesStart = true),
+            x2.copy(end = newEnd, includesEnd = true))
+        case _ =>
+          rangeBounds.head.copy(start = newStart, includesStart = true)  +:
+            rangeBounds.tail.init :+
+            rangeBounds.last.copy(end = newEnd, includesEnd = true)
```

Comment:
This is much easier to read with the match statement! I like it.

For the last case you can also do:

```scala
case head +: tail => head.copy(...) +: tail.init :+ tail.last.copy(...)
```